### PR TITLE
add support for list network area and test (#381)

### PR DESCRIPTION
**Pull request by larrytamnjong**

### CODE REVIEW

**Comment by larrytamnjong** [comment]
@marcin-krystianc

File: Consul.Test/OperatorTest.cs [hunk 1]
```diff
@@ -104,5 +104,11 @@ public async Task Operator_CreateArea()
             var response = await _client.Operator.CreateArea(check);
             Assert.NotNull(response.Response);
         }
+        [EnterpriseOnlyFact]
+        public async Task Operator_AreaList()
+        {
+            var req = await _client.Operator.AreaList();
+            Assert.NotNull(req.Response);
```
Comment (marcin-krystianc) [inline Consul.Test/OperatorTest.cs:111]: To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).

Comment (larrytamnjong) [reply Consul.Test/OperatorTest.cs:113]: Thanks, done ✅

Comment (marcin-krystianc) [reply Consul.Test/OperatorTest.cs:113]: Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.

**Review by marcin-krystianc — COMMENTED** [review]
The test are failing now with:
```
Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
            Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
        }	        }
        [EnterpriseOnlyFact]
        public async Task Operator_AreaList()
        {
            await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });

            var req = await _client.Operator.AreaList();
            Assert.NotNull(req.Response);
Member
@[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).

Contributor
Author
@[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
Thanks, done ✅

Member
@[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.

@marcin-krystianc	Reply...
        }
    }	    }
}	}
```

Maybe try using a random name every time you create an area.

**Comment by larrytamnjong** [comment]
> The test are failing now with:
> 
> ```
> Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
>             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
>         }	        }
>         [EnterpriseOnlyFact]
>         public async Task Operator_AreaList()
>         {
>             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> 
>             var req = await _client.Operator.AreaList();
>             Assert.NotNull(req.Response);
> Member
> @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> 
> Contributor
> Author
> @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> Thanks, done ✅
> 
> Member
> @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> 
> @marcin-krystianc	Reply...
>         }
>     }	    }
> }	}
> ```
> 
> Maybe try using a random name every time you create an area.

Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay now.

**Comment by marcin-krystianc** [comment]
> > The test are failing now with:
> > ```
> > Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
> >             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
> >         }	        }
> >         [EnterpriseOnlyFact]
> >         public async Task Operator_AreaList()
> >         {
> >             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> > 
> >             var req = await _client.Operator.AreaList();
> >             Assert.NotNull(req.Response);
> > Member
> > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> > To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> > 
> > Contributor
> > Author
> > @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> > Thanks, done ✅
> > 
> > Member
> > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> > Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> > 
> > @marcin-krystianc	Reply...
> >         }
> >     }	    }
> > }	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe try using a random name every time you create an area.
> 
> Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay.

Using fixed name is not ideal, I guess it is impossible to run that test twice in a row? The test should be implemented in a way, that makes it possible to run it multiple times without any issue.

**Comment by larrytamnjong** [comment]
> > > The test are failing now with:
> > > ```
> > > Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
> > >             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
> > >         }	        }
> > >         [EnterpriseOnlyFact]
> > >         public async Task Operator_AreaList()
> > >         {
> > >             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> > > 
> > >             var req = await _client.Operator.AreaList();
> > >             Assert.NotNull(req.Response);
> > > Member
> > > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> > > To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> > > 
> > > Contributor
> > > Author
> > > @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> > > Thanks, done ✅
> > > 
> > > Member
> > > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> > > Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> > > 
> > > @marcin-krystianc	Reply...
> > >         }
> > >     }	    }
> > > }	}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Maybe try using a random name every time you create an area.
> > 
> > 
> > Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay.
> 
> Using fixed name is not ideal, I guess it is impossible to run that test twice in a row? The test should be implemented in a way, that makes it possible to run it multiple times without any issue.

Thanks, @marcin-krystianc, and apologies for the back-and-forth. I’ve added a method to generate a random dc name for each test run. The tests are still in queue. Let me know if this approach works or if there's anything else you'd like me to adjust.

File: Consul.Test/OperatorTest.cs [hunk 2]
```diff
@@ -17,13 +17,21 @@
...
 
 namespace Consul.Test
 {
     public class OperatorTest : BaseFixture
     {
+        private static readonly Random Random = new Random();
+        internal static string GeneratePeerDatacenterName()
+        {
+            int randomNumber = Random.Next(100, 1000);
+            return $"ran-dc-{randomNumber}";
```
Comment (marcin-krystianc) [inline Consul.Test/OperatorTest.cs:32]: Single random number might be not enough to prevent from conflicts/duplicates. We've already a method `KVTest.GenerateTestKeyName()` which is being used in many places in tests already. It generates 16 character long random string. Please use it instead.

Comment (larrytamnjong) [reply Consul.Test/OperatorTest.cs]: @marcin-krystianc thanks, you can take another look now.

File: Consul.Test/OperatorTest.cs [hunk 3]
```diff
@@ -17,13 +17,15 @@
...
 
+using System;
 using System.Threading.Tasks;
 using Xunit;
 
 namespace Consul.Test
 {
     public class OperatorTest : BaseFixture
     {
+
```
Comment (marcin-krystianc) [inline Consul.Test/OperatorTest.cs:28]: nit: extra empty line

File: Consul.Test/OperatorTest.cs [hunk 4]
```diff
@@ -100,9 +102,20 @@ public async Task Segment_List()
...
             Assert.NotNull(response.Response);
         }
+        [EnterpriseOnlyFact]
+        public async Task Operator_AreaList()
+        {
+            var peerDataCenter = KVTest.GenerateTestKeyName();
+            await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = peerDataCenter, UseTLS = false, RetryJoin = null });
+
+            var req = await _client.Operator.AreaList();
+            Assert.NotEmpty(req.Response);
```
Comment (marcin-krystianc) [inline Consul.Test/OperatorTest.cs:118]: Just one more thing, can we test if the fields in the response are not null? 
So far we know that we can create and list areas, but it would be great if we assert that the values returned by the list operation correspond to the values used in the create operation.

File: Consul/Interfaces/IOperatorEndpoint.cs [hunk 5]
```diff
@@ -47,6 +47,7 @@ public interface IOperatorEndpoint
         Task<QueryResult<string[]>> SegmentList(CancellationToken cancellationToken = default);
         Task<WriteResult<string>> CreateArea(AreaRequest area, WriteOptions q, CancellationToken ct = default);
         Task<WriteResult<string>> CreateArea(AreaRequest area, CancellationToken ct = default);
-
+        Task<QueryResult<List<Area>>> AreaList(CancellationToken cancellationToken = default);
```
Comment (marcin-krystianc) [inline Consul/Interfaces/IOperatorEndpoint.cs:50]: We should keep the method names consistent with others. I think `AreaList` is fine, but we should rename `CreateArea` to `AreaCreate`. It would help with `Area` method discoverability (user just types Area and presses autocomplete to see available options). What do you think?

Comment (larrytamnjong) [reply Consul/Interfaces/IOperatorEndpoint.cs:50]: Makes a lot to sense ✅

File: Consul.Test/OperatorTest.cs (hunk 4 above)
Comment (larrytamnjong) [reply Consul.Test/OperatorTest.cs:121]: ✅ done

File: Consul.Test/OperatorTest.cs [hunk 6]
```diff
@@ -98,11 +101,26 @@ public async Task Segment_List()
...
+        public async Task Operator_AreaList()
+        {
+            var peerDataCenter = KVTest.GenerateTestKeyName();
+            var area = new AreaRequest { PeerDatacenter = peerDataCenter, UseTLS = false, RetryJoin = null };
+
+            await _client.Operator.AreaCreate(area);
+
+            var req = await _client.Operator.AreaList();
+            Assert.NotEmpty(req.Response);
+            Assert.Equal(area.PeerDatacenter, req.Response.First().PeerDatacenter);
```
Comment (marcin-krystianc) [inline Consul.Test/OperatorTest.cs:122]: There can be multiple items in the result list, so the first one is not necessarily the one we are looking for. You need to look for the right one, e.g.: 
```
var result = req.Response.Single(x => x.PeerDatacenter == area.peerDataCenter);
Assert.Equal(area.UseTLS, result.UseTLS);
Assert.Equal(area.RetryJoin, result.RetryJoin);  
```

File: Consul.Test/OperatorTest.cs (hunk 4 above)
Comment (marcin-krystianc) [reply Consul.Test/OperatorTest.cs:121]: Can you also try filling the `RetryJoin` field and then validating the content of it when returned?

File: Consul.Test/OperatorTest.cs (hunk 6 above)
Comment (larrytamnjong) [reply Consul.Test/OperatorTest.cs]: @marcin-krystianc I'm not sure if testing RetryJoin is sufficient with this approach ```Assert.Equal(area.RetryJoin, result.RetryJoin);``` 
or if it would be better to test each element individually in the array, like this: ```for (int i = 0; i < retryJoinAddresses.Length; i++)
    {
        Assert.Equal(retryJoinAddresses[i], createdArea.RetryJoin[i]);
    }```

Comment (marcin-krystianc) [reply Consul.Test/OperatorTest.cs]: > @marcin-krystianc I'm not sure if testing RetryJoin is sufficient with this approach `Assert.Equal(area.RetryJoin, result.RetryJoin);` or if it would be better to test each element individually in the array, like this: `for (int i = 0; i < retryJoinAddresses.Length; i++) { Assert.Equal(retryJoinAddresses[i], createdArea.RetryJoin[i]); }`

There is an overload for `IEnumerable`:
```
    public static void Equal<T>(
      IEnumerable<T> expected,
      IEnumerable<T> actual,
      IEqualityComparer<T> comparer)
```

so I guess it will do the right thing. 
You can try to define manually two lists  and see how it reacts to different item order in both lists.

Comment (larrytamnjong) [reply Consul.Test/OperatorTest.cs]: image
Tests 1 and 3 are passing, but test 2 is failing so its ok. I have pushed the changes please let me know if there's anything else to change.

**Review by marcin-krystianc — APPROVED** [review]
It's perfect now.
